[BUG] ViaFabricPlus 3.1.0 is not compatible with Nextgen 0.1.9 git-156f7e7

### LiquidBounce Branch

Nextgen

### LiquidBounce Build/Version

0.1.9 git-156f7e7

### Operating System

Windows 10

### Minecraft Version

1.20.4

### Describe the bug

When I update ViaFabricPlus from 3.0.6 to 3.1.0, it causes Minecraft to crash. After ruling out other mods, I determined that the conflict was caused by LiquidBounce.

### Steps to reproduce

Start Minecraft after updating ViaFabricPlus to 3.1.0 in an existing environment where ViaFabricPlus 3.0.6 and LiquidBounce coexist.

### Client Log

```shell
[06:28:45] [main/INFO]: Loading Minecraft 1.20.4 with Fabric Loader 0.15.7
[06:28:45] [ForkJoinPool-1-worker-4/WARN]: Mod org_cloudburstmc_netty_netty-transport-raknet uses the version 1.0.0.CR1-SNAPSHOT which isn't compatible with Loader's extended semantic version format (Could not parse version number component 'CR1'!), SemVer is recommended for reliably evaluating dependencies and prioritizing newer version
[06:28:45] [main/INFO]: Loading 93 mods:
	- fabric-api 0.96.4+1.20.4
	   |-- fabric-api-base 0.4.36+78d798af4f
	   |-- fabric-api-lookup-api-v1 1.6.49+82b1bb3e4f
	   |-- fabric-biome-api-v1 13.0.16+78d798af4f
	   |-- fabric-block-api-v1 1.0.16+3e2216cb4f
	   |-- fabric-block-view-api-v2 1.0.4+78d798af4f
	   |-- fabric-blockrenderlayer-v1 1.1.46+78d798af4f
	   |-- fabric-client-tags-api-v1 1.1.7+78d798af4f
	   |-- fabric-command-api-v1 1.2.41+f71b366f4f
	   |-- fabric-command-api-v2 2.2.20+78d798af4f
	   |-- fabric-commands-v0 0.2.58+df3654b34f
	   |-- fabric-containers-v0 0.1.86+df3654b34f
	   |-- fabric-content-registries-v0 5.0.14+78d798af4f
	   |-- fabric-convention-tags-v1 1.5.10+78d798af4f
	   |-- fabric-crash-report-info-v1 0.2.23+78d798af4f
	   |-- fabric-data-attachment-api-v1 1.1.3+b90db5744f
	   |-- fabric-data-generation-api-v1 13.2.3+5c0133444f
	   |-- fabric-dimensions-v1 2.1.61+78d798af4f
	   |-- fabric-entity-events-v1 1.6.1+09fc25014f
	   |-- fabric-events-interaction-v0 0.7.1+389931eb4f
	   |-- fabric-events-lifecycle-v0 0.2.74+df3654b34f
	   |-- fabric-game-rule-api-v1 1.0.46+78d798af4f
	   |-- fabric-item-api-v1 2.2.0+d6f2b0844f
	   |-- fabric-item-group-api-v1 4.0.25+58f8c0124f
	   |-- fabric-key-binding-api-v1 1.0.41+78d798af4f
	   |-- fabric-keybindings-v0 0.2.39+df3654b34f
	   |-- fabric-lifecycle-events-v1 2.3.0+a67ffb5d4f
	   |-- fabric-loot-api-v2 2.1.8+78d798af4f
	   |-- fabric-message-api-v1 6.0.5+78d798af4f
	   |-- fabric-mining-level-api-v1 2.1.64+78d798af4f
	   |-- fabric-model-loading-api-v1 1.0.8+78d798af4f
	   |-- fabric-models-v0 0.4.7+9386d8a74f
	   |-- fabric-networking-api-v1 3.1.7+2e5ac5484f
	   |-- fabric-object-builder-api-v1 13.0.13+080016e44f
	   |-- fabric-particles-v1 1.1.7+78d798af4f
	   |-- fabric-recipe-api-v1 2.0.20+78d798af4f
	   |-- fabric-registry-sync-v0 4.0.19+58f8c0124f
	   |-- fabric-renderer-api-v1 3.2.4+78d798af4f
	   |-- fabric-renderer-indigo 1.5.4+78d798af4f
	   |-- fabric-renderer-registries-v1 3.2.53+df3654b34f
	   |-- fabric-rendering-data-attachment-v1 0.3.42+73761d2e4f
	   |-- fabric-rendering-fluids-v1 3.1.1+e761c6694f
	   |-- fabric-rendering-v0 1.1.56+df3654b34f
	   |-- fabric-rendering-v1 3.2.0+6fd945a04f
	   |-- fabric-resource-conditions-api-v1 2.3.14+78d798af4f
	   |-- fabric-resource-loader-v0 0.11.19+58f8c0124f
	   |-- fabric-screen-api-v1 2.0.17+78d798af4f
	   |-- fabric-screen-handler-api-v1 1.3.55+78d798af4f
	   |-- fabric-sound-api-v1 1.0.17+78d798af4f
	   |-- fabric-transfer-api-v1 4.0.11+e84342304f
	   \-- fabric-transitive-access-wideners-v1 5.0.14+78d798af4f
	- fabric-language-kotlin 1.10.18+kotlin.1.9.22
	   |-- org_jetbrains_kotlin_kotlin-reflect 1.9.22
	   |-- org_jetbrains_kotlin_kotlin-stdlib 1.9.22
	   |-- org_jetbrains_kotlin_kotlin-stdlib-jdk7 1.9.22
	   |-- org_jetbrains_kotlin_kotlin-stdlib-jdk8 1.9.22
	   |-- org_jetbrains_kotlinx_atomicfu-jvm 0.23.2
	   |-- org_jetbrains_kotlinx_kotlinx-coroutines-core-jvm 1.8.0
	   |-- org_jetbrains_kotlinx_kotlinx-coroutines-jdk8 1.8.0
	   |-- org_jetbrains_kotlinx_kotlinx-datetime-jvm 0.5.0
	   |-- org_jetbrains_kotlinx_kotlinx-serialization-cbor-jvm 1.6.2
	   |-- org_jetbrains_kotlinx_kotlinx-serialization-core-jvm 1.6.2
	   \-- org_jetbrains_kotlinx_kotlinx-serialization-json-jvm 1.6.2
	- fabricloader 0.15.7
	   \-- mixinextras 0.3.5
	- java 17
	- liquidbounce 0.1.9
	- minecraft 1.20.4
	- viafabricplus 3.1.0
	   |-- com_google_code_findbugs_jsr305 3.0.2
	   |-- com_vdurmont_semver4j 3.1.0
	   |-- com_viaversion_viabackwards-common 4.10.0-24w09a-SNAPSHOT
	   |-- com_viaversion_viaversion-common 4.10.0-24w09a-SNAPSHOT
	   |-- de_florianmichael_classic4j 2.0.2
	   |-- fabric-command-api-v2 2.2.20+78d798af4f
	   |-- fabric-lifecycle-events-v1 2.3.0+a67ffb5d4f
	   |-- fabric-networking-api-v1 3.1.7+2e5ac5484f
	   |-- fabric-particles-v1 1.1.7+78d798af4f
	   |-- io_jsonwebtoken_jjwt-api 0.12.5
	   |-- io_jsonwebtoken_jjwt-gson 0.12.5
	   |-- io_jsonwebtoken_jjwt-impl 0.12.5
	   |-- net_jodah_expiringmap 0.5.10
	   |-- net_lenni0451_commons_httpclient 1.4.1
	   |-- net_lenni0451_mcping 1.4.0
	   |-- net_lenni0451_mcstructs-bedrock_forms 1.2.0
	   |-- net_lenni0451_mcstructs-bedrock_text 1.2.0
	   |-- net_lenni0451_reflect 1.3.2
	   |-- net_raphimc_minecraftauth 4.0.0-SNAPSHOT
	   |-- net_raphimc_viaaprilfools 2.0.11-SNAPSHOT
	   |-- net_raphimc_viabedrock 0.0.5-SNAPSHOT
	   |-- net_raphimc_vialegacy 2.2.22-SNAPSHOT
	   |-- net_raphimc_vialoader 2.2.13-SNAPSHOT
	   |-- org_cloudburstmc_netty_netty-transport-raknet 1.0.0.CR1-SNAPSHOT
	   |-- org_iq80_leveldb_leveldb 0.12
	   |-- org_iq80_leveldb_leveldb-api 0.12
	   |-- org_lz4_lz4-pure-java 1.8.0
	   \-- org_yaml_snakeyaml 2.2
[06:28:46] [main/INFO]: SpongePowered MIXIN Subsystem Version=0.8.5 Source=file:/H:/Games/Play2/.minecraft/libraries/net/fabricmc/sponge-mixin/0.12.5+mixin.0.8.5/sponge-mixin-0.12.5+mixin.0.8.5.jar Service=Knot/Fabric Env=CLIENT
[06:28:46] [main/INFO]: Compatibility level set to JAVA_17
[06:28:47] [main/WARN]: Error loading class: me/jellysquid/mods/sodium/client/render/chunk/compile/pipeline/BlockOcclusionCache (java.lang.ClassNotFoundException: me/jellysquid/mods/sodium/client/render/chunk/compile/pipeline/BlockOcclusionCache)
[06:28:47] [main/INFO]: Initializing MixinExtras via com.llamalad7.mixinextras.service.MixinExtrasServiceImpl(version=0.3.5).
[06:28:49] [main/WARN]: @Redirect conflict. Skipping viafabricplus.mixins.json:fixes.minecraft.entity.MixinPlayerEntity from mod viafabricplus->@Redirect::changeSpeedCalculation(Lnet/minecraft/class_1657;Lnet/minecraft/class_1291;)Z with priority 1000, already redirected by liquidbounce.mixins.json:minecraft.entity.MixinPlayerEntity from mod liquidbounce->@Redirect::injectFatigueNoSlow(Lnet/minecraft/class_1657;Lnet/minecraft/class_1291;)Z with priority 1337
[06:28:50] [main/WARN]: @Redirect conflict. Skipping viafabricplus.mixins.json:fixes.minecraft.entity.MixinPlayerEntity from mod viafabricplus->@Redirect::changeSpeedCalculation(Lnet/minecraft/class_1657;Lnet/minecraft/class_1291;)Z with priority 1000, already redirected by liquidbounce.mixins.json:minecraft.entity.MixinPlayerEntity from mod liquidbounce->@Redirect::injectFatigueNoSlow(Lnet/minecraft/class_1657;Lnet/minecraft/class_1291;)Z with priority 1337
[06:28:50] [main/ERROR]: Minecraft has crashed!
net.fabricmc.loader.impl.FormattedException: java.lang.RuntimeException: Mixin transformation of net.minecraft.class_1657 failed
	at net.fabricmc.loader.impl.FormattedException.ofLocalized(FormattedException.java:63) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:472) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) ~[fabric-loader-0.15.7.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at oolloo.jlw.Wrapper.invokeMain(Wrapper.java:71) ~[?:?]
	at oolloo.jlw.Wrapper.main(Wrapper.java:51) ~[?:?]
Caused by: java.lang.RuntimeException: Mixin transformation of net.minecraft.class_1657 failed
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:427) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:323) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:119) ~[fabric-loader-0.15.7.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
	at net.minecraft.client.main.Main.main(Main.java:178) ~[1.20.4-Fabric%200.15.7.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:470) ~[fabric-loader-0.15.7.jar:?]
	... 8 more
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:392) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:422) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:323) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:119) ~[fabric-loader-0.15.7.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
	at net.minecraft.client.main.Main.main(Main.java:178) ~[1.20.4-Fabric%200.15.7.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:470) ~[fabric-loader-0.15.7.jar:?]
	... 8 more
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector changeSpeedCalculation(Lnet/minecraft/class_1657;Lnet/minecraft/class_1291;Lcom/llamalad7/mixinextras/sugar/ref/LocalFloatRef;)Z in viafabricplus.mixins.json:fixes.minecraft.entity.MixinPlayerEntity from mod viafabricplus failed injection check, (0/1) succeeded. Scanned 1 target(s). Using refmap viafabricplus-refmap.json
	at org.spongepowered.asm.mixin.injection.struct.InjectionInfo.postInject(InjectionInfo.java:468) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at com.llamalad7.mixinextras.wrapper.InjectorWrapperImpl.doPostInject(InjectorWrapperImpl.java:79) ~[mixinextras-0.3.5-63e25e1b0e82f847.jar:?]
	at com.llamalad7.mixinextras.sugar.impl.SugarWrapperImpl.doPostInject(SugarWrapperImpl.java:84) ~[mixinextras-0.3.5-63e25e1b0e82f847.jar:?]
	at com.llamalad7.mixinextras.wrapper.WrapperInjectionInfo.postInject(WrapperInjectionInfo.java:78) ~[mixinextras-0.3.5-63e25e1b0e82f847.jar:?]
	at org.spongepowered.asm.mixin.transformer.MixinTargetContext.applyInjections(MixinTargetContext.java:1384) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyInjections(MixinApplicatorStandard.java:1062) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.applyMixin(MixinApplicatorStandard.java:402) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinApplicatorStandard.apply(MixinApplicatorStandard.java:327) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.apply(TargetClassContext.java:421) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.TargetClassContext.applyMixins(TargetClassContext.java:403) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:363) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:234) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClassBytes(MixinTransformer.java:202) ~[sponge-mixin-0.12.5+mixin.0.8.5.jar:0.12.5+mixin.0.8.5]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.getPostMixinClassByteArray(KnotClassDelegate.java:422) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.tryLoadClass(KnotClassDelegate.java:323) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:218) ~[fabric-loader-0.15.7.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:119) ~[fabric-loader-0.15.7.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:525) ~[?:?]
	at net.minecraft.client.main.Main.main(Main.java:178) ~[1.20.4-Fabric%200.15.7.jar:?]
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:470) ~[fabric-loader-0.15.7.jar:?]
	... 8 more
```


### Screenshots

![20240316062054](https://github.com/CCBlueX/LiquidBounce/assets/161578211/14e6e4cf-5a47-4c28-96cf-b016050a1b05)
![20240316062313](https://github.com/CCBlueX/LiquidBounce/assets/161578211/5550dbf5-3c78-4840-91b9-ef1f6814cadb)

--- a/build.gradle
+++ b/build.gradle
@@ -101,10 +101,7 @@ dependencies {
     // Recommended mods (on IDE)
     modRuntimeOnly "com.terraformersmc:modmenu:${project.mod_menu_version}"
     modImplementation "maven.modrinth:sodium:${project.sodium_version}"
-    modImplementation ("de.florianmichael:viafabricplus:3.0.6") {
-        exclude group: 'net.raphimc', module: 'ViaLoader'
-    }
-    implementation "net.raphimc:ViaLoader:2.2.12" // with 2.2.13 the VersionEnum disappears
+    modImplementation "de.florianmichael:viafabricplus:3.1.0"
 
     // Tests
     modImplementation 'com.github.superblaubeere27:tenacc:e3a7ada99a'

--- a/src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/entity/MixinPlayerEntity.java
+++ b/src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/entity/MixinPlayerEntity.java
@@ -19,6 +19,7 @@
 
 package net.ccbluex.liquidbounce.injection.mixins.minecraft.entity;
 
+import com.llamalad7.mixinextras.injector.ModifyExpressionValue;
 import net.ccbluex.liquidbounce.event.EventManager;
 import net.ccbluex.liquidbounce.event.events.PlayerJumpEvent;
 import net.ccbluex.liquidbounce.event.events.PlayerSafeWalkEvent;
@@ -35,11 +36,7 @@ import net.ccbluex.liquidbounce.utils.aiming.AimPlan;
 import net.ccbluex.liquidbounce.utils.aiming.Rotation;
 import net.ccbluex.liquidbounce.utils.aiming.RotationManager;
 import net.minecraft.client.MinecraftClient;
-import net.minecraft.entity.effect.StatusEffect;
-import net.minecraft.entity.effect.StatusEffects;
 import net.minecraft.entity.player.PlayerEntity;
-import net.minecraft.registry.tag.FluidTags;
-import net.minecraft.registry.tag.TagKey;
 import net.minecraft.util.Pair;
 import net.minecraft.util.math.Vec3d;
 import org.spongepowered.asm.mixin.Mixin;
@@ -122,28 +119,26 @@ public abstract class MixinPlayerEntity extends MixinLivingEntity {
         }
     }
 
-    @Redirect(method = "getBlockBreakingSpeed", at = @At(value = "INVOKE",
+    @ModifyExpressionValue(method = "getBlockBreakingSpeed", at = @At(value = "INVOKE",
             target = "Lnet/minecraft/entity/player/PlayerEntity;hasStatusEffect(Lnet/minecraft/entity/effect/StatusEffect;)Z"))
-    private boolean injectFatigueNoSlow(PlayerEntity instance, StatusEffect statusEffect) {
+    private boolean injectFatigueNoSlow(boolean original) {
         ModuleNoSlowBreak module = ModuleNoSlowBreak.INSTANCE;
-        if ((Object) this == MinecraftClient.getInstance().player &&
-                module.getEnabled() && module.getMiningFatigue() && statusEffect == StatusEffects.MINING_FATIGUE) {
+        if ((Object) this == MinecraftClient.getInstance().player && module.getEnabled() && module.getMiningFatigue()) {
             return false;
         }
 
-        return instance.hasStatusEffect(statusEffect);
+        return original;
     }
 
-    @Redirect(method = "getBlockBreakingSpeed", at = @At(value = "INVOKE",
+    @ModifyExpressionValue(method = "getBlockBreakingSpeed", at = @At(value = "INVOKE",
             target = "Lnet/minecraft/entity/player/PlayerEntity;isSubmergedIn(Lnet/minecraft/registry/tag/TagKey;)Z"))
-    private boolean injectWaterNoSlow(PlayerEntity instance, TagKey tagKey) {
+    private boolean injectWaterNoSlow(boolean original) {
         ModuleNoSlowBreak module = ModuleNoSlowBreak.INSTANCE;
-        if ((Object) this == MinecraftClient.getInstance().player && tagKey == FluidTags.WATER &&
-                module.getEnabled() && module.getWater()) {
+        if ((Object) this == MinecraftClient.getInstance().player && module.getEnabled() && module.getWater()) {
             return false;
         }
 
-        return instance.isSubmergedIn(tagKey);
+        return original;
     }
 
     @Redirect(method = "getBlockBreakingSpeed", at = @At(value = "INVOKE",

--- a/src/main/java/net/ccbluex/liquidbounce/utils/client/vfp/Vfp306Compatibility.java
+++ b/src/main/java/net/ccbluex/liquidbounce/utils/client/vfp/Vfp306Compatibility.java
@@ -1,0 +1,112 @@
+/*
+ * This file is part of LiquidBounce (https://github.com/CCBlueX/LiquidBounce)
+ *
+ * Copyright (c) 2024 CCBlueX
+ *
+ * LiquidBounce is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * LiquidBounce is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with LiquidBounce. If not, see <https://www.gnu.org/licenses/>.
+ *
+ *
+ */
+
+package net.ccbluex.liquidbounce.utils.client.vfp;
+
+import net.ccbluex.liquidbounce.LiquidBounce;
+import net.ccbluex.liquidbounce.utils.client.ClientProtocolVersion;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.List;
+
+/**
+ * Compatibility layer for ViaFabricPlus
+ *
+ * DO NOT CALL ANY OF THESE METHODS WITHOUT CHECKING IF VIAFABRICPLUS IS LOADED
+ */
+public enum Vfp306Compatibility {
+
+    INSTANCE;
+
+    public ClientProtocolVersion unsafeGetProtocolVersion() {
+        try {
+            Class<?> protocolHackClass = Class.forName("de.florianmichael.viafabricplus.protocolhack.ProtocolHack");
+            Method getTargetVersionMethod = protocolHackClass.getMethod("getTargetVersion");
+            Object versionEnum = getTargetVersionMethod.invoke(null);
+            return new ClientProtocolVersion((String) versionEnum.getClass().getMethod("getName").invoke(versionEnum),
+                    (Integer) versionEnum.getClass().getMethod("getVersion").invoke(versionEnum));
+        } catch (Throwable throwable) {
+            LiquidBounce.INSTANCE.getLogger().error("Failed to get protocol version", throwable);
+            return null;
+        }
+    }
+
+    public ClientProtocolVersion[] unsafeGetProtocolVersions() {
+        try {
+            Class<?> versionEnumClass = Class.forName("net.raphimc.vialoader.util.VersionEnum");
+            Field sortedVersionsField = versionEnumClass.getField("SORTED_VERSIONS");
+            List sortedVersions = (List) sortedVersionsField.get(null);
+            return (ClientProtocolVersion[]) sortedVersions.stream()
+                    .map(version -> {
+                        try {
+                            return new ClientProtocolVersion((String) version.getClass().getMethod("getName").invoke(version),
+                                    (Integer) version.getClass().getMethod("getVersion").invoke(version));
+                        } catch (IllegalAccessException e) {
+                            throw new RuntimeException(e);
+                        } catch (InvocationTargetException e) {
+                            throw new RuntimeException(e);
+                        } catch (NoSuchMethodException e) {
+                            throw new RuntimeException(e);
+                        }
+                    })
+                    .toArray(ClientProtocolVersion[]::new);
+        } catch (Throwable throwable) {
+            LiquidBounce.INSTANCE.getLogger().error("Failed to get protocol versions", throwable);
+            return new ClientProtocolVersion[0];
+        }
+    }
+
+    public void unsafeSelectProtocolVersion(int protocolId) {
+        try {
+            Class<?> versionEnumClass = Class.forName("net.raphimc.vialoader.util.VersionEnum");
+            Method fromProtocolIdMethod = versionEnumClass.getMethod("fromProtocolId", int.class);
+            Object versionEnum = fromProtocolIdMethod.invoke(null, protocolId);
+
+            if (versionEnum == null) {
+                throw new IllegalStateException("Protocol version " + protocolId + " not found");
+            }
+
+            Class<?> protocolHackClass = Class.forName("de.florianmichael.viafabricplus.protocolhack.ProtocolHack");
+            Method setTargetVersionMethod = protocolHackClass.getMethod("setTargetVersion", versionEnumClass);
+            setTargetVersionMethod.invoke(null, versionEnum);
+        } catch (Throwable throwable) {
+            LiquidBounce.INSTANCE.getLogger().error("Failed to select protocol version", throwable);
+        }
+    }
+
+    public boolean isOldCombat() {
+        try {
+            Class<?> protocolHackClass = Class.forName("de.florianmichael.viafabricplus.protocolhack.ProtocolHack");
+            Method getTargetVersionMethod = protocolHackClass.getMethod("getTargetVersion");
+            Object versionEnum = getTargetVersionMethod.invoke(null);
+            Method isOlderThanOrEqualToMethod = versionEnum.getClass().getMethod("isOlderThanOrEqualTo", versionEnum.getClass());
+            Object result = isOlderThanOrEqualToMethod.invoke(versionEnum,
+                    Class.forName("net.raphimc.vialoader.util.VersionEnum").getField("r1_8").get(null));
+            return (boolean) result;
+        } catch (Throwable throwable) {
+            LiquidBounce.INSTANCE.getLogger().error("Failed to check if old combat", throwable);
+            return false;
+        }
+    }
+
+}

--- a/src/main/java/net/ccbluex/liquidbounce/utils/client/vfp/VfpCompatibility.java
+++ b/src/main/java/net/ccbluex/liquidbounce/utils/client/vfp/VfpCompatibility.java
@@ -21,14 +21,15 @@
 
 package net.ccbluex.liquidbounce.utils.client.vfp;
 
-import de.florianmichael.viafabricplus.protocolhack.ProtocolHack;
+import com.viaversion.viaversion.api.protocol.version.ProtocolVersion;
+import de.florianmichael.viafabricplus.protocoltranslator.ProtocolTranslator;
 import de.florianmichael.viafabricplus.screen.base.ProtocolSelectionScreen;
 import de.florianmichael.viafabricplus.settings.impl.VisualSettings;
 import net.ccbluex.liquidbounce.LiquidBounce;
-import net.ccbluex.liquidbounce.utils.client.ProtocolVersion;
+import net.ccbluex.liquidbounce.utils.client.ClientProtocolVersion;
 import net.minecraft.client.MinecraftClient;
 import net.minecraft.client.gui.screen.TitleScreen;
-import net.raphimc.vialoader.util.VersionEnum;
+import org.apache.commons.lang3.ArrayUtils;
 
 /**
  * Compatibility layer for ViaFabricPlus
@@ -51,25 +52,28 @@ public enum VfpCompatibility {
         }
     }
 
-    public ProtocolVersion unsafeGetProtocolVersion() {
+    public ClientProtocolVersion unsafeGetProtocolVersion() {
         try {
-            VersionEnum version = ProtocolHack.getTargetVersion();
-            return new ProtocolVersion(version.getProtocol().getName(), version.getProtocol().getVersion());
+            ProtocolVersion version = ProtocolTranslator.getTargetVersion();
+            return new ClientProtocolVersion(version.getName(), version.getVersion());
         } catch (Throwable throwable) {
             LiquidBounce.INSTANCE.getLogger().error("Failed to get protocol version", throwable);
             return null;
         }
     }
 
-    public ProtocolVersion[] unsafeGetProtocolVersions() {
+    public ClientProtocolVersion[] unsafeGetProtocolVersions() {
         try {
-            return VersionEnum.SORTED_VERSIONS
+            var protocols = ProtocolVersion.getProtocols()
                     .stream()
-                    .map(version -> new ProtocolVersion(version.getProtocol().getName(), version.getProtocol().getVersion()))
-                    .toArray(ProtocolVersion[]::new);
+                    .map(version -> new ClientProtocolVersion(version.getName(), version.getVersion()))
+                    .toArray(ClientProtocolVersion[]::new);
+
+            ArrayUtils.reverse(protocols);
+            return protocols;
         } catch (Throwable throwable) {
             LiquidBounce.INSTANCE.getLogger().error("Failed to get protocol versions", throwable);
-            return new ProtocolVersion[0];
+            return new ClientProtocolVersion[0];
         }
     }
 
@@ -88,12 +92,12 @@ public enum VfpCompatibility {
 
     public void unsafeSelectProtocolVersion(int protocolId) {
         try {
-            VersionEnum version = VersionEnum.fromProtocolId(protocolId);
-            if (version == null) {
-                throw new IllegalStateException("Protocol version " + protocolId + " not found");
+            if (!ProtocolVersion.isRegistered(protocolId)) {
+                throw new IllegalArgumentException("Protocol version is not registered");
             }
 
-            ProtocolHack.setTargetVersion(version);
+            ProtocolVersion version = ProtocolVersion.getProtocol(protocolId);
+            ProtocolTranslator.setTargetVersion(version);
         } catch (Throwable throwable) {
             LiquidBounce.INSTANCE.getLogger().error("Failed to select protocol version", throwable);
         }
@@ -101,10 +105,10 @@ public enum VfpCompatibility {
 
     public boolean isOldCombat() {
         try {
-            var version = ProtocolHack.getTargetVersion();
+            var version = ProtocolTranslator.getTargetVersion();
 
             // Check if the version is older or equal than 1.8
-            return version.isOlderThanOrEqualTo(VersionEnum.r1_8);
+            return version.olderThanOrEqualTo(ProtocolVersion.v1_8);
         } catch (Throwable throwable) {
             LiquidBounce.INSTANCE.getLogger().error("Failed to check if old combat", throwable);
             return false;

--- a/src/main/java/net/ccbluex/liquidbounce/utils/client/vfp/VfpCompatibility.java
+++ b/src/main/java/net/ccbluex/liquidbounce/utils/client/vfp/VfpCompatibility.java
@@ -22,6 +22,7 @@
 package net.ccbluex.liquidbounce.utils.client.vfp;
 
 import com.viaversion.viaversion.api.protocol.version.ProtocolVersion;
+import com.viaversion.viaversion.api.protocol.version.VersionType;
 import de.florianmichael.viafabricplus.protocoltranslator.ProtocolTranslator;
 import de.florianmichael.viafabricplus.screen.base.ProtocolSelectionScreen;
 import de.florianmichael.viafabricplus.settings.impl.VisualSettings;
@@ -66,6 +67,7 @@ public enum VfpCompatibility {
         try {
             var protocols = ProtocolVersion.getProtocols()
                     .stream()
+                    .filter(version -> version.getVersionType() == VersionType.RELEASE)
                     .map(version -> new ClientProtocolVersion(version.getName(), version.getVersion()))
                     .toArray(ClientProtocolVersion[]::new);
 

--- a/src/main/kotlin/net/ccbluex/liquidbounce/utils/client/ProtocolUtil.kt
+++ b/src/main/kotlin/net/ccbluex/liquidbounce/utils/client/ProtocolUtil.kt
@@ -20,6 +20,7 @@
  */
 package net.ccbluex.liquidbounce.utils.client
 
+import net.ccbluex.liquidbounce.utils.client.vfp.Vfp306Compatibility
 import net.ccbluex.liquidbounce.utils.client.vfp.VfpCompatibility
 import net.minecraft.SharedConstants
 
@@ -34,6 +35,11 @@ val hasProtocolHack = runCatching {
     true
 }.getOrDefault(false)
 
+val hasProtocolTranslator = runCatching {
+    Class.forName("de.florianmichael.viafabricplus.protocoltranslator.ProtocolTranslator")
+    true
+}.getOrDefault(false)
+
 val hasVisualSettings = runCatching {
     Class.forName("de.florianmichael.viafabricplus.settings.impl.VisualSettings")
     true
@@ -42,43 +48,49 @@ val hasVisualSettings = runCatching {
 /**
  * Both 1.20.3 and 1.20.4 use protocol 765, so we can use this as a default
  */
-val defaultProtocolVersion = ProtocolVersion(SharedConstants.getGameVersion().name,
+val defaultProtocolVersion = ClientProtocolVersion(SharedConstants.getGameVersion().name,
     SharedConstants.getGameVersion().protocolVersion)
 
-val protocolVersion: ProtocolVersion
+val protocolVersion: ClientProtocolVersion
     get() = runCatching {
         // Check if the ViaFabricPlus mod is loaded - prevents from causing too many exceptions
-        if (!hasProtocolHack) {
+        if (hasProtocolTranslator) {
+            return@runCatching VfpCompatibility.INSTANCE.unsafeGetProtocolVersion()
+        } else if (hasProtocolHack) {
+            return@runCatching Vfp306Compatibility.INSTANCE.unsafeGetProtocolVersion()
+        } else {
             return@runCatching defaultProtocolVersion
         }
-
-        VfpCompatibility.INSTANCE.unsafeGetProtocolVersion()
     }.onFailure {
         logger.error("Failed to get protocol version", it)
     }.getOrDefault(defaultProtocolVersion)
 
-val protocolVersions: Array<ProtocolVersion>
+val protocolVersions: Array<ClientProtocolVersion>
     get() = runCatching {
         // Check if the ViaFabricPlus mod is loaded - prevents from causing too many exceptions
-        if (!hasProtocolHack) {
+        if (hasProtocolTranslator) {
+            return@runCatching VfpCompatibility.INSTANCE.unsafeGetProtocolVersions()
+        } else if (hasProtocolHack) {
+            return@runCatching Vfp306Compatibility.INSTANCE.unsafeGetProtocolVersions()
+        } else {
             return@runCatching arrayOf(defaultProtocolVersion)
         }
-
-        VfpCompatibility.INSTANCE.unsafeGetProtocolVersions()
     }.onFailure {
         logger.error("Failed to get protocol version", it)
     }.getOrDefault(arrayOf(defaultProtocolVersion))
 
-data class ProtocolVersion(val name: String, val version: Int)
+data class ClientProtocolVersion(val name: String, val version: Int)
 
 val isOldCombat: Boolean
     get() = runCatching {
         // Check if the ViaFabricPlus mod is loaded - prevents from causing too many exceptions
-        if (!hasProtocolHack) {
+        if (hasProtocolTranslator) {
+            return@runCatching VfpCompatibility.INSTANCE.isOldCombat
+        } else if (hasProtocolHack) {
+            return@runCatching Vfp306Compatibility.INSTANCE.isOldCombat
+        } else {
             return@runCatching false
         }
-
-        VfpCompatibility.INSTANCE.isOldCombat
     }.onFailure {
         logger.error("Failed to check if the server is using old combat", it)
     }.getOrDefault(false)
@@ -87,11 +99,13 @@ val isOldCombat: Boolean
 
 fun selectProtocolVersion(protocolId: Int) {
     // Check if the ViaFabricPlus mod is loaded - prevents from causing too many exceptions
-    if (!usesViaFabricPlus) {
+    if (hasProtocolTranslator) {
+        VfpCompatibility.INSTANCE.unsafeSelectProtocolVersion(protocolId)
+    } else if (hasProtocolHack) {
+        Vfp306Compatibility.INSTANCE.unsafeSelectProtocolVersion(protocolId)
+    } else {
         error("ViaFabricPlus is not loaded")
     }
-
-    VfpCompatibility.INSTANCE.unsafeSelectProtocolVersion(protocolId)
 }
 
 fun openVfpProtocolSelection() {

